File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        fast, slow ,prev = head, head, None
        while fast and fast.next:
            fast = fast.next.next
            temp = slow.next
            slow.next = prev
            prev = slow
            slow = temp

        if fast: fast = slow.next
        else: fast = slow
        slow = prev

        while fast and slow:
            if fast.val != slow.val: 
                return False
            fast = fast.next
            slow = slow.next
        
        return True
        

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    let fast = head, slow = head, prev = null;
    while(fast != null && fast.next != null) {
        fast = fast.next.next;
        let temp = slow.next;
        slow.next = prev;
        prev = slow;
        slow = temp;
    }

    if(fast != null) fast = slow.next;
    else fast = slow;
    slow = prev;

    while(fast != null && slow != null) {
        if(fast.val != slow.val) return false;
        fast = fast.next;
        slow = slow.next;
    }

    return true;
};